### PR TITLE
Fuse `InlineLet` and `UpdateLet` into `RelationCSE`

### PR DESCRIPTION
**Pull request by ggevay**
Up to now, it used to be a rule that `RelationCSE` must always be followed by `InlineLet`. This commit makes `RelationCSE` automatically call `InlineLet`, which has several advantages:
 - It is impossible to forget calling `InlineLet` after `RelationCSE`.
 - It is impossible to accidentally insert something between them.
 - The optimizer trace is a bit cleaner: Before this change, `RelationCSE` almost always made a big, messy diff, and then `InlineLet` often reverted the plan to be the same or similar as before `RelationCSE`. Now we see the changes made by `RelationCSE` + `InlineLet` as just one diff. This is often an empty diff, and even when it is not an empty diff, it's a smaller diff than before.

I also fused `UpdateLet` into `RelationCSE`. Note that only 2 of the 3 instances of `RelationCSE` had `UpdateLet` after it. I'm not sure why the 3rd one (in `logical_cleanup_pass`) didn't have that. Now I also fused `UpdateLet` into `RelationCSE`, so all 3 instances of `RelationCSE` now involve `UpdateLet`. Is this ok, @frankmcsherry?

The random reviewer is @vmarcos.

### Motivation

   * This PR refactors existing code.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
  - [ ] The already existing tests cover it: There shouldn't be any explain plan changes. (The only test changes are those that show plans at different optimizer steps.)
  - [ ] I also ran the optimizer trace tool manually.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - None

### CODE REVIEW

File: src/transform/src/lib.rs
Comment (aalexandrov) [inline src/transform/src/lib.rs:400]: nit: I think you can remove this comment now, as with the proposed setup the lack of explicit inlining entry in the pipeline might cause confusion.

File: src/transform/src/cse/relation_cse.rs
Comment (aalexandrov) [inline src/transform/src/cse/relation_cse.rs:63]: Hmm, we now have a situation where the need to trace a call is conditional based on the client..

I'm not super fond of having to introduce dedicated methods such as `transform_without_trace` just to disable. I think that we should have it the other way around - `transform` should not trace the call, and instead the `Transform` trait should implement a default `trace_transform` method that adds the necessary decoration around the `self.transform()` call. This will also address [@frankmcsherry's concern  from #16157](https://github.com/MaterializeInc/materialize/pull/16157#discussion_r1028140532).

The top-level client call should then decide whether it wants to be tracing or not. I'll do a follow-up PR to address this after this and #16157 land.

**Review by aalexandrov — APPROVED** [review]
I have some remarks around the way this has turned out in terms of ad-hoc `transform_without_trace` (see commend below), but I would prefer to address this in a separate PR.

File: src/transform/src/lib.rs
Comment (ggevay) [reply src/transform/src/lib.rs]: Yes!

**Comment by ggevay** [comment]
Thanks for the comments, @aalexandrov!

**Review by vmarcos — APPROVED** [review]
LGTM! @ggevay Thank you for the quick chat about this PR.

**Review by frankmcsherry — APPROVED** [review]
The idea works great for me, and I'll defer to @aalexandrov on the most idiomatic approach to writing the tracing code.

**Pull request merged**
